tests/Model/LazyBSONArrayTest.php

```diff
@@ -0,0 +1,289 @@
+<?php
+
+namespace MongoDB\Tests\Model;
+
+use Generator;
+use MongoDB\BSON\PackedArray;
+use MongoDB\Exception\InvalidArgumentException;
+use MongoDB\Model\LazyBSONArray;
+use MongoDB\Model\LazyBSONDocument;
+use MongoDB\Tests\TestCase;
+use stdClass;
+
+use function iterator_to_array;
+use function json_encode;
+use function serialize;
+use function unserialize;
+
+use const JSON_THROW_ON_ERROR;
+
+class LazyBSONArrayTest extends TestCase
+{
+    public static function provideTestArray(): Generator
+    {
+        yield 'array' => [
+            new LazyBSONArray([
+                'bar',
+                new LazyBSONDocument(['bar' => 'baz']),
+                new LazyBSONArray([0, 1, 2]),
+            ]),
+        ];
+
+        yield 'packedArray' => [
+            new LazyBSONArray(PackedArray::fromPHP([
+                'bar',
+                ['bar' => 'baz'],
+                [0, 1, 2],
+            ])),
+        ];
+    }
+
+    public function testConstructWithoutArgument(): void
+    {
+        $instance = new LazyBSONArray();
+        $this->assertSame([], iterator_to_array($instance));
+    }
+
+    public function testConstructWithWrongType(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        new LazyBSONArray('foo');
+    }
+
+    public function testConstructWithArrayUsesLiteralValues(): void
+    {
+        $array = new LazyBSONArray([
+            (object) ['bar' => 'baz'],
+            ['bar' => 'baz'],
+            [0, 1, 2],
+        ]);
+
+        $this->assertInstanceOf(stdClass::class, $array[0]);
+        $this->assertIsArray($array[1]);
+        $this->assertIsArray($array[2]);
+    }
+
+    public function testClone(): void
```

Comment:
This test makes me realize that we don't actually have any tests for `MongoDB\recursive_copy()`. I suppose the best test there would be to recursively walk the original and cloned values and then assert all the all objects are equal but not identical.